Geometry_Engine/Modify/SplitAtPoints.cs

```diff
@@ -212,111 +212,113 @@ public static List<PolyCurve> SplitAtPoints(this PolyCurve curve, List<Point> po
                 return new List<PolyCurve> { curve.DeepClone() };
 
             List<PolyCurve> result = new List<PolyCurve>();
-            List<ICurve> tmpResult = new List<ICurve>();
-            List<Point> subPoints = new List<Point>();
-            List<Point> onCurvePoints = new List<Point>();
 
-            foreach (Point p in points)
+            List<ICurve> subParts = curve.ISubParts().ToList(); //Subparts of PolyCurve
+            if (subParts.Count == 1)    //Single SubPart
             {
-                if (p.IsOnCurve(curve, tolerance))
-                    onCurvePoints.Add(p);
+                foreach (ICurve splitSegment in SplitAtPoints(subParts[0] as dynamic, points, tolerance))   //Simply call the appropriate split at method
+                {
+                    result.Add(new PolyCurve { Curves = new List<ICurve> { splitSegment } });   //And wrap split segments into polycurves
+                }
+                return result;  
             }
 
-            onCurvePoints = onCurvePoints.CullDuplicates(tolerance);
-
-            if (onCurvePoints.Count == 0)
-                return new List<PolyCurve> { curve.DeepClone() };
-
-            onCurvePoints = onCurvePoints.SortAlongCurve(curve);
-
-            foreach (ICurve crv in curve.SubParts())
+            List<Point> nonDuplicatePoints = points.CullDuplicates(tolerance);
+            PolyCurve prev = null;  //Polycurve to collect parts across segments
+            double sqTol = tolerance * tolerance;
+            for (int k = 0; k < subParts.Count; k++)
             {
-                if (crv is Arc)
+                ICurve crv = subParts[k];
+                List<Point> ptsOnSegment = nonDuplicatePoints.Where(x => x.IIsOnCurve(crv, tolerance)).ToList(); //Get points on the segment
```

Comment:
From a brief look at the other `SplitAtPoints` methods, `IsOnCurve` is called explicitly inside each of them. Which means that this call is likely to be redundant.

Reply:
It kind of is redundant, yes, but there to check as a pre-check if a split will be made for the particular segment, if ptsOnSegment count is non-zero. 

Can refactor this slightly though to send in all points, and then, if a single item is returned, check if start and/or endpoints are within tolerance of any of the points instead. Might potentially be cleaner..

Reply:
Yeah, I think it may be cleaner, plus also potentially faster, assuming that point operations are lighter than duplicate `IsOnCurve` (run once to get `ptsOnSegment` and then called individually inside `SplitAtPoints`)

Reply:
fixed now, with final refactoring